x509/revocation/ocsp/ocsp.go

```diff
@@ -1,173 +1,364 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package ocsp parses OCSP responses as specified in RFC 2560. OCSP responses
+// are signed messages attesting to the validity of a certificate for a small
+// period of time. This is used to manage revocation for X.509 certificates.
 package ocsp
 
 import (
 	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	_ "crypto/sha1"
+	_ "crypto/sha256"
+	_ "crypto/sha512"
 	"encoding/asn1"
-	"encoding/json"
 	"errors"
+	"fmt"
 	"math/big"
+	"strconv"
 	"time"
 
 	"github.com/zmap/zcrypto/x509"
 	"github.com/zmap/zcrypto/x509/pkix"
 	"github.com/zmap/zcrypto/x509/revocation/crl"
 )
 
+var idPKIXOCSPBasic = asn1.ObjectIdentifier([]int{1, 3, 6, 1, 5, 5, 7, 48, 1, 1})
+
 // ResponseStatus contains the result of an OCSP request. See
 // https://tools.ietf.org/html/rfc6960#section-2.3
 type ResponseStatus int
 
-// Success - OCSP Responder signals that response is successful
-const Success ResponseStatus = 0
-
-var responseStatusNames map[ResponseStatus]string
-var hashOIDs map[crypto.Hash]asn1.ObjectIdentifier
-
-func init() {
-	responseStatusNames = make(map[ResponseStatus]string)
-	responseStatusNames[0] = "success"
-	responseStatusNames[1] = "malformedRequest"
-	responseStatusNames[2] = "internalError"
-	responseStatusNames[3] = "tryLater"
-	responseStatusNames[5] = "sigRequired"
-	responseStatusNames[6] = "unauthorized"
-
-	hashOIDs = make(map[crypto.Hash]asn1.ObjectIdentifier)
-	hashOIDs[crypto.SHA1] = asn1.ObjectIdentifier([]int{1, 3, 14, 3, 2, 26})
-	hashOIDs[crypto.SHA256] = asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 1})
-	hashOIDs[crypto.SHA384] = asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 2})
-	hashOIDs[crypto.SHA512] = asn1.ObjectIdentifier([]int{2, 16, 840, 1, 101, 3, 4, 2, 3})
-}
+const (
+	Success       ResponseStatus = 0
+	Malformed     ResponseStatus = 1
+	InternalError ResponseStatus = 2
+	TryLater      ResponseStatus = 3
+	// Status code four is unused in OCSP. See
+	// https://tools.ietf.org/html/rfc6960#section-4.2.1
+	SignatureRequired ResponseStatus = 5
+	Unauthorized      ResponseStatus = 6
+)
 
-// MarshalJSON implements the json.Marshler interface
-func (code *ResponseStatus) MarshalJSON() ([]byte, error) {
-	aux := struct {
-		Value int    `json:"value"`
-		Name  string `json:"name"`
-	}{
-		Value: int(*code),
-		Name:  code.String(),
+func (r ResponseStatus) String() string {
+	switch r {
+	case Success:
+		return "success"
+	case Malformed:
+		return "malformed"
+	case InternalError:
+		return "internal error"
+	case TryLater:
+		return "try later"
+	case SignatureRequired:
+		return "signature required"
+	case Unauthorized:
+		return "unauthorized"
+	default:
+		return "unknown OCSP status: " + strconv.Itoa(int(r))
 	}
-	return json.Marshal(&aux)
```

Comment:
We vendored and edited it to add JSON support.

Reply:
@dadrian I can try to restore JSON support if it's really needed.
what is the scenario for JSON in OCSP?

Reply:
I believe it was done for data collection purposes for bulk revocation. I don't know that revocation ever came to fruition outside of some Censys prototypes, so it's probably fine to drop.